weasyl/searchtag.py

```diff
@@ -160,24 +160,33 @@ def select_list(map_table, targetids):
     return dict(list(db.execute(q)))
 
 
-@region.cache_on_arguments()
-def _get_or_create(name):
-    tag = d.engine.scalar(
-        'INSERT INTO searchtag (title) VALUES (%(name)s) ON CONFLICT (title) DO NOTHING RETURNING tagid',
-        name=name)
+@region.cache_multi_on_arguments()
+def _get_or_create(*names: str) -> list[int]:
+    names_list = list(names)
+
+    d.engine.execute('''
+        INSERT INTO searchtag (title)
+        SELECT title
+        FROM UNNEST(%(names)s::text[]) AS title
+        ON CONFLICT (title) DO NOTHING
+    ''', names=names_list)
 
-    if tag is not None:
-        return tag
+    result = d.engine.execute(
+        'SELECT tagid FROM searchtag WHERE title = ANY (%(names)s)',
+        names=names_list)
 
-    return d.engine.scalar(
-        'SELECT tagid FROM searchtag WHERE title = %(name)s',
-        name=name)
+    return result.scalars().all()
 
 
 def get_or_create(name):
     return _get_or_create(d.get_search_tag(name))
 
 
+def get_or_create_many(names: list[str]) -> list[int]:
+    normalized_names = map(d.get_search_tag, names)
```

Comment:
I know it’s a little unsatisfying because this is named similar to `get_or_create`, but given that this doesn’t deduplicate or remove empty tags and the value it’s passed is already normalized, how about this patch:

```patch
diff --git a/weasyl/blocktag.py b/weasyl/blocktag.py
index 11930804..b9dea48d 100644
--- a/weasyl/blocktag.py
+++ b/weasyl/blocktag.py
@@ -79,7 +79,7 @@ def insert(userid: int, tags: str, rating: int):
 
     profile.check_user_rating_allowed(userid, rating)
 
-    parsed_tags = list(searchtag.parse_tags(tags))
+    parsed_tags = searchtag.parse_tags(tags)
 
     d.engine.execute('''
         INSERT INTO blocktag (userid, tagid, rating)
diff --git a/weasyl/searchtag.py b/weasyl/searchtag.py
index 0275ef0d..71443a46 100644
--- a/weasyl/searchtag.py
+++ b/weasyl/searchtag.py
@@ -5,6 +5,7 @@ import secrets
 import time
 from dataclasses import dataclass
 from struct import Struct
+from typing import Iterable
 
 import sqlalchemy as sa
 from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
@@ -182,8 +183,7 @@ def get_or_create(name):
     return _get_or_create(d.get_search_tag(name))
 
 
-def get_or_create_many(names: list[str]) -> list[int]:
-    normalized_names = map(d.get_search_tag, names)
+def get_or_create_many(normalized_names: Iterable[str]) -> list[int]:
     return _get_or_create(*normalized_names)
 
 
```

Reply:
That's a good idea! Applied that now.